src/transformers/debug_utils.py

```diff
@@ -0,0 +1,325 @@
+
+
+class DebugUnderflowOverflow:
+    """
+    This debug class helps detect and understand where the model starts getting very large or very small, and more
+    importantly ``nan`` or ``inf`` weight and activation elements.
+
+    There are 2 working modes:
+
+    1. Underflow/overflow detection (default)
+    2. Specific batch absolute min/max tracing without detection
+
+    Mode 1: Underflow/overflow detection
+
+    To activate the underflow/overflow detection, initialize the object with the model ::
+
+        debug_overflow = DebugUnderflowOverflow(model)
+
+    then run the training as normal and if ``nan`` or ``inf`` gets detected in at least one of the weight, input or
+    output elements this module will throw an exception and will print ``max_frames_to_save`` frames that lead to this
+    event, each frame reporting
+
+    1. the fully qualified module name plus the class name whose ``forward`` was run
+    2. the absolute min and max value of all elements for each module weights, and the inputs and output
+
+    For example, here is the header and the last few frames in detection report for ``google/mt5-small`` run in fp16 mixed precision ::
+
+        Detected inf/nan during batch_number=0
+        Last 21 forward frames:
+        abs min  abs max  metadata
+        [...]
+                          encoder.block.2.layer.1.DenseReluDense.wi_0 Linear
+        2.17e-07 4.50e+00 weight
+        1.79e-06 4.65e+00 input[0]
+        2.68e-06 3.70e+01 output
+                          encoder.block.2.layer.1.DenseReluDense.wi_1 Linear
+        8.08e-07 2.66e+01 weight
+        1.79e-06 4.65e+00 input[0]
+        1.27e-04 2.37e+02 output
+                          encoder.block.2.layer.1.DenseReluDense.wo Linear
+        1.01e-06 6.44e+00 weight
+        0.00e+00 9.74e+03 input[0]
+        3.18e-04 6.27e+04 output
+                          encoder.block.2.layer.1.DenseReluDense T5DenseGatedGeluDense
+        1.79e-06 4.65e+00 input[0]
+        3.18e-04 6.27e+04 output
+                          encoder.block.2.layer.1.dropout Dropout
+        3.18e-04 6.27e+04 input[0]
+        0.00e+00      inf output
+
+    You can see here, that ``T5DenseGatedGeluDense.forward`` resulted in output activations, whose absolute max value
+    was around 62.7K, which is very close to fp16's top limit of 64K. In the next frame we have ``Dropout`` which
+    renormalizes the weights, after it zeroed some of the elements, which pushes the absolute max value to more than
+    64K, and we get an overlow.
+
+    As you can see it's the previous frames that we need to look into when the numbers start going into very large for
+    fp16 numbers.
+
+    The tracking is done in a forward hook, which gets invoked immediately after ``forward`` has completed.
+
+    By default the last 21 frames are printed. You can change the default to adjust for your needs. For example ::
+
+        debug_overflow = DebugUnderflowOverflow(model, max_frames_to_save=100)
+
+
```

Comment:
@sgugger, how could I add some kind of small separator here? I tried `----` but sphinx complains. Thank you!

Reply:
or alternatively making the `Mode 1` / `Model 2` into small headers - not sure how to do that in docstring.

Reply:
I have no idea how to do this inside a docstring, which means IMHO that you may be trying to put too much in a docstring? We can expand more in `docs/source/debugging.rst`

Reply:
The ReST standard has an `<hr>` equivalent using `----` but sphinx fails on it.

That's OK, I'm sure it will evolve. This is the initial pass.

Since this tool can be used outside of `transformers` I think ideally it should have all the essential info in its docstring.

So I suggest just leaving it as is for now.